test/recursive_wrapper_move_only_test.cpp

```diff
@@ -0,0 +1,69 @@
+/*=============================================================================
+  Copyright (c) 2019 Nikita Kniazev
+
+  Distributed under the Boost Software License, Version 1.0. (See accompanying
+  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+=============================================================================*/
+
+// Check that move-only types can be used with recursive_wrapper
+
+#include <boost/config.hpp>
+#ifdef BOOST_NO_CXX11_RVALUE_REFERENCES
+// variant does not support move emulation
+int main() {}
+#else
+#include "boost/core/lightweight_test.hpp"
+#include <boost/variant.hpp>
+#include <boost/move/utility_core.hpp>
+
+struct move_only_type
+{
+private:
+    BOOST_MOVABLE_BUT_NOT_COPYABLE(move_only_type)
+
+public:
+    explicit move_only_type(int value) BOOST_NOEXCEPT
+      : value_(value)
+    {}
+
+    // Move ctor
+    move_only_type(BOOST_RV_REF(move_only_type) x) BOOST_NOEXCEPT
+      : value_(x.value_)
+    {}
+
+    // Move assign
+    move_only_type& operator=(BOOST_RV_REF(move_only_type) x) BOOST_NOEXCEPT
+    {
+        value_ = x.value_;
+        return *this;
+    }
+
+    int value_;
+};
+
+int main()
+{
+    // TODO: variant does not have inplace construction yet
+    move_only_type a_(0), b_(1);
+    boost::variant<
+        boost::recursive_wrapper<int>,
+        boost::recursive_wrapper<move_only_type>
+    > a(boost::move(a_)), b(boost::move(b_)), c(3);
+    BOOST_TEST(!a.empty());
+    BOOST_TEST(!b.empty());
+    a = boost::move(b);
+    BOOST_TEST(!a.empty());
+    BOOST_TEST_EQ(boost::get<move_only_type>(a).value_, 1);
+    // check that we can reuse a moved out variant with move assignment
+    move_only_type c_(2);
+    b = boost::move(c_); // TODO: no emplace yet
+    BOOST_TEST(!b.empty());
+    BOOST_TEST_EQ(boost::get<move_only_type>(b).value_, 2);
+    // check that recursive variant advertises its valueless after beging moved out
+    BOOST_TEST_EQ(boost::get<int>(c), 3);
+    c = boost::move(b);
+    BOOST_TEST( b.empty());
+    BOOST_TEST_EQ(boost::get<move_only_type>(c).value_, 2);
+    return boost::report_errors();
```

Comment:
Add tests for throwing move assignments and recursive wrapper. Make sure that if the type `A` has a throwing default constructor and an always throwing move assignment then the following code works:
```
variant<A, recursive_wrapper<int>> a(A()), rec(1);
try {
    rec = std::move(a);
    Assert(false);
} catch (...) { BOOST_TEST(!rec.empty()); }
```

Also make sure that 
```
variant<A, recursive_wrapper<int>> a(A()), rec(1), rec2(2);
rec2 = std::move(rec)
try {
    rec = std::move(a);
    Assert(false);
} catch (...) { BOOST_TEST(rec.empty()); }
rec = std::move(rec2)
```
is fine.

Reply:
From what I see variant will never use throwing move even if copy can throw too (variant also does not use move assignment, as it constructs a temporary variant and swaps the content). In the second example I am do not understand why `rec` expected to be empty after failed move.

Reply:
You are right, `rec` could be anything. Just make sure that it destructs well after the `throw` in move assignment.

Reply:
I am not sure how can I check if it destructs "well". Is not the first example does it already?

Reply:
Ping

Reply:
Everything looks great. I'll merge the PR after the 1.70 release (I'm not comfortable with merging it now, without a few weeks of usage in develop)

Reply:
Ok, thanks for the clarification.